### PR TITLE
PERF: custom ops for RangeIndex.[all|any|__contains__]

### PR DESCRIPTION
**Pull request by topper-123**
- [x] xref #26565
- [x] tests added / passed
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [ ] whatsnew entry

Follow-up to #26565. Make ``RangeIndex._data`` be created in fewer cases.

Performance examples (but the larger gain is from memory savings by not creating the ``_data`` array):

```python
>>>  rng = pd.RangeIndex(-5_000_000, 0, 6)
>>> %timeit rng.all()
789 µs ± 5.24 µs per loop  # master
216 ns ± 0.957 ns per loop  # this PR
>>> %timeit rng.any()
763 µs ± 6.08 µs per loop  # master
124 ns ± 9.15 ns per loop  # this PR
>>> %timeit (-2 in rng)
380 ns ± 1.85 ns per loop  # master
689 ns ± 44.7 ns per loop  # this PR, slightly slower
```

### CODE REVIEW

File: pandas/core/indexes/range.py
Comment (topper-123) [inline pandas/core/indexes/range.py]: A Python int is needed, else Python iterates over the range, making the operation very slow.

File: pandas/core/common.py
Comment (jreback) [inline pandas/core/common.py]: put in dtypes.cast but i don’t think u need this as we already have ensure_platform_int

Comment (topper-123) [reply pandas/core/common.py]: ``ensure_platform_int`` returns a numpy int(32|64), so won't work here:

```python
>>> rng = range(1, 1_000_000)
>>> 999_999 in rng
133 ns ± 2.46 ns per loop
>>> np.int64(999_999) in rng
306 ms ± 66.3 ms per loop
```

Comment (jreback) [reply pandas/core/common.py]: ok in any event move it to collocate with the ensure _ routines

File: pandas/core/indexes/range.py
Comment (jschendel) [inline pandas/core/indexes/range.py]: this can be reduced to `return 0 not in self._range`

Comment (topper-123) [reply pandas/core/indexes/range.py]: Yeah, that's better. Changed.

File: pandas/core/dtypes/cast.py
Comment (jreback) [inline pandas/core/dtypes/cast.py]: I gave the wrong direction, can you put in core/dtypes/common (right below the other ensure_*)

and can you add a doc-string (I know its trivial)

the typing seems odd here, shouldn't this be Union[int, np.integer] ?

File: pandas/core/indexes/range.py
Comment (jreback) [inline pandas/core/indexes/range.py]: import directly

File: pandas/core/indexes/range.py
Comment (jreback) [inline pandas/core/indexes/range.py]: can you type

**Review by jreback — CHANGES_REQUESTED** [review]
small comments

File: pandas/core/dtypes/cast.py
Comment (topper-123) [reply pandas/core/dtypes/cast.py]: I'm not quite sure. In principle a float or a np.floating and possibly others could for some subvalues pass the test (e.g. if value is 1.0). But ok, passing a float would be a code smell, so just as well make the programmer see passing those as typing errors.

File: pandas/core/indexes/range.py
Comment (topper-123) [reply pandas/core/indexes/range.py]: I will need to wait for this until #26581 is merged, as mypy doesn't understand/see attributes defined in ``__new__`` or ``_simple_new`` . If that one is merged, I can go through this again wrt. typing.

**Comment by topper-123** [comment]
Comments addressed. The typing issue must wait until #26581 is merged for mypy to see _range.

**Comment by jreback** [comment]
can you add the issue number onto the one you already have in Performance section; merge master as well.  ping on green.

**Pull request merged**
